storage/retention.go

```diff
@@ -98,149 +91,41 @@ func (s *retentionEnforcer) run() {
 
 // expireData runs a delete operation on the storage engine.
 //
-// Any series data that (1) belongs to a bucket in the provided map and
+// Any series data that (1) belongs to a bucket in the provided list and
 // (2) falls outside the bucket's indicated retention period will be deleted.
-func (s *retentionEnforcer) expireData(rpByBucketID map[platform.ID]time.Duration, now time.Time) error {
-	_, logEnd := logger.NewOperation(s.logger, "Data deletion", "data_deletion")
+func (s *retentionEnforcer) expireData(buckets []*platform.Bucket, now time.Time) error {
+	logger, logEnd := logger.NewOperation(s.logger, "Data deletion", "data_deletion")
 	defer logEnd()
 
-	ctx, cancel := context.WithTimeout(context.Background(), engineAPITimeout)
-	defer cancel()
-	cur, err := s.Engine.CreateSeriesCursor(ctx, SeriesCursorRequest{}, nil)
-	if err != nil {
-		return err
-	}
-	defer cur.Close()
-
-	var mu sync.Mutex
-	badMSketch := make(map[string]struct{})          // Badly formatted measurements.
-	missingBSketch := make(map[platform.ID]struct{}) // Missing buckets.
-
-	var seriesDeleted uint64 // Number of series where a delete is attempted.
-	var seriesSkipped uint64 // Number of series that were skipped from delete.
-
-	fn := func(name []byte, tags models.Tags) (int64, int64, bool) {
-		if len(name) != platform.IDLength {
-			mu.Lock()
-			badMSketch[string(name)] = struct{}{}
-			mu.Unlock()
-			atomic.AddUint64(&seriesSkipped, 1)
-			return 0, 0, false
-
+	for _, b := range buckets {
+		if b.RetentionPeriod == 0 {
+			continue
 		}
 
-		var n [16]byte
-		copy(n[:], name)
-		_, bucketID := tsdb.DecodeName(n)
-
-		retentionPeriod, ok := rpByBucketID[bucketID]
-		if !ok {
-			mu.Lock()
-			missingBSketch[bucketID] = struct{}{}
-			mu.Unlock()
-			atomic.AddUint64(&seriesSkipped, 1)
-			return 0, 0, false
-		}
-		if retentionPeriod == 0 {
-			return 0, 0, false
+		max := now.Add(-b.RetentionPeriod).UnixNano()
+		err := s.Engine.DeleteBucketRange(b.OrganizationID, b.ID, math.MinInt64, max)
+		if err != nil {
+			// TODO(jeff): metrics?
```

Comment:
I don't see where `err` can be returned here? Maybe change the API so it doesn't return an `error`? 

Then you could move the failed metrics error down from here https://github.com/influxdata/influxdb/pull/11364/files#diff-b633e12e59f6e30d9f2f96ccc8a301e2R84 into this branch?

We should track failed deletions segmented by org, bucket, which is what https://github.com/influxdata/influxdb/pull/11364/files#diff-b633e12e59f6e30d9f2f96ccc8a301e2R89 tracks.